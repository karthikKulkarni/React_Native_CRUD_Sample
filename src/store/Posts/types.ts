export type Post = {
  userId: number;
  id?: number;
  title: string;
  body: string;
};

//Get All Posts
export type AllPostsInput = {
  id: number;
};

export type AllPostsData = Post[];

export type AllPostsType =
  | { type: 'POSTS_REQUEST'; input?: AllPostsInput }
  | { type: 'POSTS_SUCCESS'; posts: AllPostsData }
  | { type: 'POSTS_FAILURE'; error: string };

//Add a post
export type AddPostInput = {
  post: Post;
};

export type AddPostData = {
  id: number;
  post: Post;
};

export type AddPostsType =
  | { type: 'ADD_POST_REQUEST'; input: AddPostInput }
  | { type: 'ADD_POST_SUCCESS'; newPost: AddPostData }
  | { type: 'ADD_POST_FAILURE'; error: string };

//Update a post

export type UpdatePostInput = {
  postBody: Post;
  postId: number;
};

export type UpdatedPostData = {
  id: number;
  post: Post;
};

export type UpdatePostType =
  | { type: 'UPDATE_POST_REQUEST'; updatePost: UpdatePostInput }
  | { type: 'UPDATE_POST_SUCCESS'; updatedPost: UpdatedPostData }
  | { type: 'UPDATE_POST_FAILURE'; error: string };

//Delete a post

export type DeletePostInput = {
  postId: number;
};

export type DeletedPostData = {
  post: Post;
};

export type DeletePostType =
  | { type: 'DELETE_POST_REQUEST'; deletePost: DeletePostInput }
  | { type: 'DELETE_POST_SUCCESS'; deletedPost: DeletedPostData }
  | { type: 'DELETE_POST_FAILURE'; error: string };

//Posts Types
export type PostTypes = AllPostsType | AddPostsType | UpdatePostType | DeletePostType;
