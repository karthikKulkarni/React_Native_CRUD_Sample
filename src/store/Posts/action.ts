import {
  Post,
  AllPostsType,
  AllPostsInput,
  AllPostsData,
  AddPostsType,
  AddPostInput,
  AddPostData,
  UpdatePostInput,
  UpdatePostData,
  UpdatePostType,
  DeletePostInput,
  DeletePostType,
  DeletedPostData,
} from './types';

//Get all posts
export function allPostsRequest(input?: AllPostsInput): AllPostsType {
  return { type: 'POSTS_REQUEST', input };
}

export function allPostsSuccess(posts: Post[]): AllPostsType {
  return { type: 'POSTS_SUCCESS', posts };
}

export function allPostsFailure(error: string): AllPostsType {
  return { type: 'POSTS_FAILURE', error };
}

//Add new posts
export function addPostsRequest(post: Post): AddPostsType {
  return { type: 'ADD_POST_REQUEST', input: { post } };
}

export function addPostsSuccess(newPost: AddPostData): AddPostsType {
  return { type: 'ADD_POST_SUCCESS', newPost };
}

export function addPostsFailure(error?: string) {
  return { type: 'ADD_POST_FAILURE', error };
}

//Update post
export function updatePostRequest(updatePost: UpdatePostInput): UpdatePostType {
  return { type: 'UPDATE_POST_REQUEST', updatePost };
}

export function updatePostsSuccess(updatedPost: UpdatePostData): UpdatePostType {
  return { type: 'UPDATE_POST_SUCCESS', updatedPost };
}

export function updatePostsFailure(error: string): UpdatePostType {
  return { type: 'UPDATE_POST_FAILURE', error };
}

//Delete a post
export function deletePostRequest(deletePost: DeletePostInput): DeletePostType {
  return { type: 'DELETE_POST_REQUEST', deletePost };
}

export function deletedPostsSuccess(deletedPost: DeletedPostData): DeletePostType {
  return { type: 'DELETE_POST_SUCCESS', deletedPost };
}

export function deletePostsFailure(error: string): DeletePostType {
  return { type: 'DELETE_POST_FAILURE', error };
}
