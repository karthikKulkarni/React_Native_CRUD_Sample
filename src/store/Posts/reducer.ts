import {
  Post,
  AllPostsType,
  AllPostsData,
  AddPostData,
  AddPostsType,
  UpdatePostType,
  UpdatedPostData,
  DeletePostType,
  DeletedPostData,
  PostTypes,
} from './types';
import { StandardApiState } from '..';
import { createEmptyState } from '../util';
import _ from 'lodash';

//Get Posts Reducer
const initialData: AllPostsData = [];
const initialState = createEmptyState(initialData);

export function allPostsReducer(state = initialState, action: AllPostsType): StandardApiState<AllPostsData> {
  switch (action.type) {
    case 'POSTS_REQUEST':
      return { ...state, isLoading: true };
    case 'POSTS_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        dateFetched: new Date(),
        data: [...action.posts], // This implementation is incorrect, something fishy
      };
    case 'POSTS_FAILURE':
      return { ...state, isLoading: false, isError: true, errorMessage: action.error };
    default:
      return state;
  }
}

//Add Post Reducer
const addPostInitData: AddPostData = {
  id: 0,
  post: {
    userId: 0,
    title: '',
    body: '',
  },
};
const addPostInitState = createEmptyState(addPostInitData);

export function addPostsReducer(
  state = addPostInitState,
  action: AddPostsType,
): StandardApiState<AddPostData> {
  switch (action.type) {
    case 'ADD_POST_REQUEST':
      return { ...state, isLoading: true };
    case 'ADD_POST_SUCCESS':
      return { ...state, isLoading: false, isError: false, dateFetched: new Date(), data: action.newPost };
    case 'ADD_POST_FAILURE':
      return { ...state, isLoading: false, isError: true, errorMessage: action.error };
    default:
      return state;
  }
}

//Update Post Reducer
const updatePostInitData: UpdatedPostData = {
  id: 0,
  post: {
    userId: 0,
    title: '',
    body: '',
  },
};
const updatePostInitState = createEmptyState(updatePostInitData);

export function updatePostsReducer(
  state = updatePostInitState,
  action: UpdatePostType,
): StandardApiState<UpdatedPostData> {
  switch (action.type) {
    case 'UPDATE_POST_REQUEST':
      return { ...state, isLoading: true };
    case 'UPDATE_POST_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        dateFetched: new Date(),
        data: action.updatedPost,
      };
    case 'UPDATE_POST_FAILURE':
      return { ...state, isLoading: false, isError: true, errorMessage: action.error };
    default:
      return state;
  }
}

//Update Post Reducer
const deletePostInitData: DeletedPostData = {
  post: {
    userId: 0,
    title: '',
    body: '',
  },
};
const deletePostInitState = createEmptyState(deletePostInitData);

export function deletePostsReducer(
  state = deletePostInitState,
  action: DeletePostType,
): StandardApiState<DeletedPostData> {
  switch (action.type) {
    case 'DELETE_POST_REQUEST':
      return { ...state, isLoading: true };
    case 'DELETE_POST_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        dateFetched: new Date(),
        data: action.deletedPost,
      };
    case 'DELETE_POST_FAILURE':
      return { ...state, isLoading: false, isError: true, errorMessage: action.error };
    default:
      return state;
  }
}

//Custom posts list

const initPost: Post[] = [];
export function postsReducer(state = initPost, action: PostTypes): Post[] {
  switch (action.type) {
    case 'POSTS_SUCCESS':
      return [...action.posts];

    case 'ADD_POST_SUCCESS':
      const newPost: any = action.newPost;
      console.log('newPost', newPost);
      newPost.id = _.get(action, 'newPost.id', 0);
      return [...state, newPost];

    case 'UPDATE_POST_SUCCESS':
      const updatedPost = _.get(action, 'updatedPost');

      const updatedList: Post[] = state.map((item: Post) =>
        item.id === updatedPost.id ? updatedPost.post : item,
      );
      return updatedList;

    case 'DELETE_POST_SUCCESS':
    case 'DELETE_POST_FAILURE':
      console.log('action', action);
      const index = _.findIndex(state, { id: _.get(action, 'deletedPost.id') });

      const updatedListDelete: Post[] = state.filter((item, i) => i !== index);
      return updatedListDelete;

    default:
      return state;
  }
}
