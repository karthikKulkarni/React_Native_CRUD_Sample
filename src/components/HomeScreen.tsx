import React, { Component } from 'react';
import * as _ from 'lodash';
import { View, Text, FlatList, TouchableOpacity, Button, Alert } from 'react-native';
import { connect } from 'react-redux';
import { GlobalState, StandardApiState } from '../store';
import {
  allPostsRequest,
  addPostsRequest,
  updatePostRequest,
  deletePostRequest,
} from '../store/Posts/action';
import { Post, UpdatePostInput, DeletePostInput, AllPostsData } from '../store/Posts/types';

export interface Props {
  allPostsRequest?: typeof allPostsRequest;
  rawPosts?: StandardApiState<AllPostsData>;
  addPostsRequest?: typeof addPostsRequest;
  updatePostRequest?: typeof updatePostRequest;
  deletePostRequest?: typeof deletePostRequest;

  posts: Post[];
}

export class __HomeScreen extends Component<Props> {
  componentDidMount() {
    this.props.allPostsRequest && this.props.allPostsRequest();
  }

  createPost = () => {
    const newPost: Post = { title: 'foo', body: 'bar', userId: 1 };
    this.props.addPostsRequest && this.props.addPostsRequest(newPost);
  };

  updatePost = () => {
    const updatePost: UpdatePostInput = { postBody: { title: 'Ufoo', body: 'Ubar', userId: 1 }, postId: 5 };
    this.props.updatePostRequest && this.props.updatePostRequest(updatePost);
  };

  deletePost = (item: Post) => {
    const deletePost: DeletePostInput = { postId: item.id ? item.id : 0 };
    this.props.deletePostRequest && this.props.deletePostRequest(deletePost);
  };

  renderItem = ({ item }: { item: Post }) => {
    return (
      <TouchableOpacity onPress={() => this.deletePost(item)}>
        <View style={{ width: '100%', flexDirection: 'column' }}>
          <Text style={{ width: '100%', backgroundColor: '#E6E6E6' }}>{item.title}</Text>
          <Text style={{ width: '100%', backgroundColor: '#F7F7F7' }}>{item.body}</Text>
        </View>
      </TouchableOpacity>
    );
  };

  renderEmptyComponent = () => {
    const { rawPosts } = this.props;

    return (
      <View style={{ alignItems: 'center', justifyContent: 'center', marginTop: 100 }}>
        {rawPosts ? (
          <Text
            style={{
              width: '100%',
              backgroundColor: rawPosts.isLoading ? 'orange' : 'red',
              textAlign: 'center',
            }}
          >
            {rawPosts.isLoading ? 'Loading...' : rawPosts.errorMessage}
          </Text>
        ) : (
          <Text> No Posts to show</Text>
        )}
      </View>
    );
  };

  render() {
    const { posts } = this.props;

    return (
      <View style={{ flex: 1, margin: 16 }} testID={'Homescreen_container'}>
        <View style={{ flexDirection: 'row', justifyContent: 'space-evenly' }}>
          <Button onPress={this.createPost} title={'Create'} testID={'Homescreen_Create_Btn'} />
          <Button onPress={this.updatePost} title={'Update'} />
        </View>
        <FlatList data={posts} renderItem={this.renderItem} ListEmptyComponent={this.renderEmptyComponent} />
      </View>
    );
  }
}

const mapStateToProps = (state: GlobalState) => ({
  rawPosts: state.rawPosts,
  rawNewPost: state.rawNewPost,
  rawUpdatePost: state.rawUpdatedPost,
  rawDeletedPost: state.rawDeletedPost,
  posts: state.posts,
});

export default connect(mapStateToProps, {
  allPostsRequest,
  addPostsRequest,
  updatePostRequest,
  deletePostRequest,
})(__HomeScreen);
