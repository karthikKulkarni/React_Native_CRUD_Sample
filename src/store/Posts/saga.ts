import { call, put, takeLatest } from 'redux-saga/effects';
import { GetPostsApi } from '../../service/GetPostsAPI';
import * as actions from './action';
import { AddPostsType } from './types';

export function* getPostsRequest(api: GetPostsApi) {
  try {
    const response = yield call(api.getAllPosts);
    const result = response.data;
    if (response.ok) {
      yield put(actions.allPostsSuccess(result));
    } else {
      yield put(actions.allPostsFailure(response.problem));
    }
  } catch (error) {
    yield put(actions.allPostsFailure(error));
  }
}

export function* addPostsRequest(api: GetPostsApi, action: any) {
  try {
    const response = yield call(api.addNewPost, action.input);
    const result = response.data;
    console.log('response addPostsRequest', action.input.post);
    if (response.ok) {
      yield put(actions.addPostsSuccess(result));
    } else {
      yield put(actions.addPostsSuccess({ id: 1010, ...action.input.post }));
      // yield put(actions.addPostsFailure(response.problem));
    }
  } catch (error) {
    console.log('ERROR', error);
    yield put(actions.addPostsFailure(error));
  }
}

export function* updatePostRequest(api: GetPostsApi, action: any) {
  const { postBody, postId } = action.updatePost;
  try {
    const response = yield call(api.updatePost, postBody, postId);
    const result = response.data;
    console.log('response', response);
    if (response.ok) {
      yield put(actions.updatePostsSuccess(result));
    } else {
      yield put(actions.updatePostsFailure(response.problem));
    }
  } catch (error) {
    yield put(actions.updatePostsFailure(error));
  }
}

export function* deletePostRequest(api: GetPostsApi, action: any) {
  const { postId } = action.deletePost;
  try {
    const response = yield call(api.deletePost, postId);
    const result = response.data;
    console.log('response', response);
    if (response.ok) {
      yield put(actions.deletedPostsSuccess(result));
    } else {
      yield put(actions.deletePostsFailure(response.problem));
    }
  } catch (error) {
    yield put(actions.deletePostsFailure(error));
  }
}
