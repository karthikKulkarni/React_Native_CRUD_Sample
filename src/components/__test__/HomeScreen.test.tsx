import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import { __HomeScreen } from '../HomeScreen';
import { Post } from '../../store/Posts/types';

describe('Home Screen', () => {
  const posts: Post[] = [{ userId: 0, id: 2, title: 'Foo', body: 'Boo' }];
  const props = {
    posts,
  };

  it('should render correctly', () => {
    const wrapper = shallow(<__HomeScreen {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should check if Homescreen_Create_Btn exists', () => {
    const wrapper = shallow(<__HomeScreen {...props} />);

    const createBtn = wrapper.findWhere(node => node.prop('testID') === 'Homescreen_Create_Btn');

    expect(createBtn).toExist();
    expect(createBtn.first().props().title).toEqual('Create');
  });
});
