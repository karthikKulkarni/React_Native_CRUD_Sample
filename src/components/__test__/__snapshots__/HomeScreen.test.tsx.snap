// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 16,
    }
  }
  testID="Homescreen_container"
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-evenly",
      }
    }
  >
    <Button
      onPress={[Function]}
      testID="Homescreen_Create_Btn"
      title="Create"
    />
    <Button
      onPress={[Function]}
      title="Update"
    />
  </View>
  <FlatList
    ListEmptyComponent={[Function]}
    data={
      Array [
        Object {
          "body": "Boo",
          "id": 2,
          "title": "Foo",
          "userId": 0,
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
